src/hidimstat/ensemble_clustered_inference.py

```diff
@@ -1,45 +1,330 @@
 import numpy as np
+from sklearn.base import clone
+from sklearn.utils import resample
 from joblib import Parallel, delayed
 
-from .clustered_inference import clustered_inference
-from .multi_sample_split import aggregate_medians, aggregate_quantiles
+from hidimstat.desparsified_lasso import (
+    desparsified_lasso,
+    desparsified_lasso_pvalue,
+    desparsified_group_lasso_pvalue,
+)
+from hidimstat.stat_tools import aggregate_quantiles
 
 
-def _ensembling(
-    list_beta_hat,
-    list_pval,
-    list_pval_corr,
-    list_one_minus_pval,
-    list_one_minus_pval_corr,
-    method="quantiles",
-    gamma_min=0.2,
+def _subsampling(n_samples, train_size, groups=None, seed=0):
+    """
+    Random subsampling for statistical inference.
+
+    Parameters
+    ----------
+    n_samples : int
+        Total number of samples in the dataset.
+    train_size : float
+        Fraction of samples to include in the training set (between 0 and 1).
+    groups : ndarray, shape (n_samples,), optional (default=None)
+        Group labels for samples. If provided, subsampling is done
+        at group level.
+    seed : int, optional (default=0)
+        Random seed for reproducibility.
+
+    Returns
+    -------
+    train_index : ndarray
+        Indices of selected samples for training.
+    """
+    index_row = np.arange(n_samples) if groups is None else np.unique(groups)
+    train_index = resample(
+        index_row,
+        n_samples=int(len(index_row) * train_size),
+        replace=False,
+        random_state=seed,
+    )
+    if groups is not None:
+        train_index = np.arange(n_samples)[np.isin(groups, train_index)]
+    return train_index
+
+
+def _rescale_beta(beta_hat, n_features, ward):
```

Comment:
what this function does is not primary a rescaling but a broadcasting of cluster-level statistics toward feature-level statistics.

Reply:
done